primitives/utils/src/metrics.rs

```diff
@@ -0,0 +1,147 @@
+
+use lazy_static::lazy_static;
+use prometheus::{
+    Registry, Error as PrometheusError, Opts,
+    core::{
+        GenericGaugeVec, AtomicU64,
+        GenericGauge,
+	}
+};
+use std::time::Instant;
+use std::convert::TryFrom;
+use parking_lot::Mutex;
+
+use crate::channels::{ Sender, Receiver, OnFullStrategy, channel_with_strategy };
+
+lazy_static! {
```

Comment:
We have tried very hard to not have any global variables within our metric system.

Why are global metric definitions hard to work with:

- First of all they represent global state, thus all pitfalls of global variables apply.

- Second of all let's say one wants to initialize two things that are using these, each reporting to different metrics (e.g. differentiated by label). This is impossible with global metric definitions. More specifically in this use case, let's say we want to run two tokio runtimes. We would like to have a counter for each.

Reply:
While I have a strong opinion on global state, I do see the benefit of global statics for short term analysis of unbounded channels given that it is behind the `metered` flag.

Reply:
While I generally see the argument of global state, I would like to counter argue that:
1. this means that you'd have to pass around the Prometheus registry everywhere, often times the specific Thing needs to go super deep into (just look at where `IMPORT BLOCK` or the channels are being used) the code base, pass around that reference. Like, there is a reason why also the Prometheus library assumes many of their examples are basically using `lazy_static!`. It is a reasonable trade-off
2. Your second argument is mostly philosphical. These are super specific gauges and metrics, yes, they should be addapted and appended when we change the system. Should we have two tokio runtimes (though as the tokio-threads vs system-threads counter shows you, we clearly do not even have a clean story of _using one tokio runtime_), we probably want to have a second counter or make these `Vec`s that use the builders-name. I do not assume this code won't change and evolve and more and other counters should be added. Not seeing how that flexibility is inhibited by having global variables.

The one argument I'd see that some of these gauges are so specific they shouldn't be _located_ here. But in particular the `tick` that is currently needed for performance reasons, makes it icky for the reasons stated in 1.

Maybe I should also clarify that while I see it is as goal to not have global state in general, much of system makes a lot of assumptions around that. In particular, for the time being, it will not be possible to run two full clients within the same instance – and that is okay for now.